src/nvim/api/tabpage.c

```diff
@@ -122,6 +122,45 @@ Window nvim_tabpage_get_win(Tabpage tabpage, Error *err)
   abort();
 }
 
+/// Sets the current window in a tabpage
+///
+/// @param tabpage  Tabpage handle, or 0 for current tabpage
+/// @param win Window handle, must already belong to {tabpage}
+/// @param[out] err Error details, if any
+void nvim_tabpage_set_win(Tabpage tabpage, Window win, Error *err)
+  FUNC_API_SINCE(12)
+{
+  tabpage_T *tp = find_tab_by_handle(tabpage, err);
+  if (!tp) {
+    return;
+  }
+
+  win_T *wp = find_window_by_handle(win, err);
+  if (!wp) {
+    return;
+  }
+
+  tabpage_T *win_tp = win_find_tabpage(wp);
+
+  if (win_tp != tp) {
+    api_set_error(err, kErrorTypeException, "Window does not belong to tabpage %d", tp->handle);
+    return;
+  }
+
+  if (tp == curtab) {
+    try_start();
+    goto_tabpage_win(win_tp, wp);
```

Comment:
As `wp_tp == tp == curtab`, `win_enter(wp, true)` could work?

Reply:
Ah this is a very good point, thanks!

Reply:
The doc comment says that autocmds may close the window immediately so the caller must check `win_valid(wp)` - does this mean I should ensure the window is valid before calling, or after?

Reply:
Not necessary here, as `wp != curwin` would handle that (`goto_tabpage_win()` would also be affected).

I guess the theoretical concern is the autocmd closing `wp` and opening a new window that happens to be allocated in the same storage as `wp`, but IIRC that's not possible from autocmds as they're not freed immediately.

Reply:
Gotcha, that makes sense. Thanks!